File: formula-java/src/main/java/org/formula/parse/tree/AnyUntilNodeExpression.java
package org.formula.parse.tree;

public class AnyUntilNodeExpression implements NodeExpression {
    private final String closeSequence;
    private final String escapeSequence;

    public static AnyUntilNodeExpression of(String closeSequence, String escapeSequence) {
        return new AnyUntilNodeExpression(closeSequence, escapeSequence);
    }

    public static AnyUntilNodeExpression of(String closeSequence) {
        return new AnyUntilNodeExpression(closeSequence, null);
    }

    @Override
    public <T> Node<T> chainTo(Node<T> root) {
        return root.add(new AnyUntilNode<>(closeSequence, escapeSequence));
    }

    private AnyUntilNodeExpression(String closeSequence, String escapeSequence) {
        this.closeSequence = closeSequence;
        this.escapeSequence = escapeSequence;
    }
}

File: formula-cs/Formula/TokenTree/AnyUntilNodeExpression.cs
namespace Formula.TokenTree;

public class AnyUntilNodeExpression : INodeExpression
{
    private readonly string _closeSequence;
    private readonly string? _escapeSequence;

    public static AnyUntilNodeExpression Of(string closeSequence, string? escapeSequence = null) {
        return new AnyUntilNodeExpression(closeSequence, escapeSequence);
    }

    public Node<T> ChainTo<T>(Node<T> root)
    {
        return root.Add(new AnyUntilNode<T>(_closeSequence, _escapeSequence));
    }

    private AnyUntilNodeExpression(string closeSequence, string? escapeSequence = null) {
        _closeSequence = closeSequence;
        _escapeSequence = escapeSequence;
    }
}